java/HealthVault Mobile/HVSample/src/main/java/com/microsoft/hsg/android/hvsample/WeightActivity.java

```diff
@@ -0,0 +1,162 @@
+import android.view.Window;
+import android.widget.AdapterView;
+import android.widget.ArrayAdapter;
+import android.widget.Button;
+import android.widget.EditText;
+import android.widget.ListView;
+import android.widget.Spinner;
+import android.widget.TextView;
+import android.widget.Toast;
+import android.widget.AdapterView.OnItemSelectedListener;
+
+public class WeightActivity extends Activity {
+
+	private HealthVaultApp service;
+    private HealthVaultClient hvClient;
+    private Record currentRecord;
+    
+    @Override
+    public void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.weight);
+        service = HealthVaultApp.getInstance();
+        hvClient = new HealthVaultClient();
+        
+        Button weightsBtn = (Button) findViewById(R.id.addWeight);
+        final EditText editText = (EditText) findViewById(R.id.weightInput);
+		
+        weightsBtn.setOnClickListener(new View.OnClickListener() {
+            public void onClick(View view) {
+            	if (service.isAppConnected()) {
+                    try {
+                        putWeight(editText.getText().toString());
+                    }
+                    catch (Exception e) {
+                        Toast.makeText(WeightActivity.this, "Please enter a weight!", Toast.LENGTH_SHORT).show();
+                    }
+            	}
+            }
+        });
+
+        setTitle("Weight Sample");
+    }
+    
+	@Override
+	protected void onStart() {
+		super.onStart();
+		hvClient.start();
+	}
+    
+    @Override
+    protected void onResume()
+    {
+        super.onResume();
+        currentRecord = HealthVaultApp.getInstance().getCurrentRecord();
+        getWeights();
+    }
+    
+    @Override
+	protected void onStop() {
+    	hvClient.stop();
+		super.onStop();
+	}
+
+    @SuppressWarnings("unchecked")
+    private void getWeights()
+    {
+    	hvClient.asyncRequest(
+    			currentRecord.getThingsAsync(ThingRequestGroup2.thingTypeQuery(Weight.ThingType)),
+    			new WeightCallback(WeightCallback.RenderWeights));
+    }
+    
+    private void putWeight(String value) {
+        final Thing2 thing = new Thing2();
+		thing.setData(new Weight(Double.parseDouble(value)));
+        // thing.setData(new Weight(-10));
+		hvClient.asyncRequest(
+				currentRecord.putThingAsync(thing),
+				new WeightCallback(WeightCallback.PutWeights));
+    }
+    
+    private void renderWeights(List<Thing2> things) {
+    	List<String> weights = new ArrayList<String>();
+        int count = 0;
+
+        for(Thing2 thing : things) {
+        	Weight w = (Weight)thing.getData();
+            if (count < 1)
+            {
+                TextView lastWeight = (TextView) findViewById(R.id.lasWeight);
+                lastWeight.setText(String.valueOf(w.getValue().getKg()));
+                count++;
+            }
+        	weights.add(String.valueOf(w.getWhen().getDate().getM() + "/" +
```

Comment:
For readability I think string.format might be a better approach here. Perf concerns are really an issue and it will help make this more readable.